recirq/optimize/mpg.py

```
@@ -0,0 +1,390 @@
+
+    # stats
+    history_max = -np.inf
+
+    while num_iter < max_iterations:
+        # get samples from the current policy to evaluate
+        z = np.random.randn(batch_size, n)
+        new_xs = sigma * z + current_x
+
+        if total_evals + batch_size > max_evaluations:
+            message = "Reached maximum number of evaluations."
+            break
+
+        # Evaluate points
+        res.xs_iters.append(new_xs)
+        new_ys = [f(x) for x in new_xs]
+        res.ys_iters.append(new_ys)
+        total_evals += batch_size
+        known_xs.extend(new_xs)
+        known_ys.extend(new_ys)
+
+        # Save function value
+        if save_func_vals:
+            res.func_vals.append(f(current_x))
+            res.x_iters.append(np.copy(current_x))
+            res.fun = res.func_vals[-1]
+
+        # current sampling radius (maximal)
+        max_radius = 0
+        for x in new_xs:
+            if np.linalg.norm(x - current_x) > max_radius:
+                max_radius = np.linalg.norm(x - current_x)
+
+        reward = [-y for y in new_ys]
+
+        # warmup steps control whether to use the model to estimate the f
+        if num_iter >= warmup_steps:
+            # Determine points to use to build model
+            model_xs = []
+            model_ys = []
+            # safer way without the `SVD` not converging
+            try:
+                for x, y in zip(known_xs, known_ys):
+                    if np.linalg.norm(x - current_x) < radius_coeff * max_radius:
+                        model_xs.append(x)
+                        model_ys.append(y)
+                model = _get_quadratic_model(model_xs, model_ys, x)
+
+                # get samples (from model)
+                z = np.random.randn(batch_size_model, n)
+                new_xs = sigma * z + current_x
+
+                # use the model for prediction
+                new_ys = model.predict(new_xs - current_x)
+                reward = [-y for y in new_ys]
+            except:
```

this is too broad of an exception catch. Please only catch errors you anticipate so this doesn't silently swallow unrelated, "real" errors

---

make it narrow